src/librustdoc/html/render/print_item.rs

```diff
@@ -1148,17 +1148,23 @@ fn item_enum(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, e: &clean::
             render_attributes_in_pre(w, it, "");
             write!(
                 w,
-                "{}enum {}{}{}",
+                "{}enum {}{}",
                 it.visibility.print_with_space(it.item_id, cx),
                 it.name.unwrap(),
                 e.generics.print(cx),
-                print_where_clause(&e.generics, cx, 0, true),
             );
+            let before_where = w.len();
+            write!(w, "{}", print_where_clause(&e.generics, cx, 0, true),);
+            // If there wasn't a `where` clause, we add a whitespace.
+            if before_where == w.len() {
```

Comment:
Determining whether there's a `where` clause based on the length of the emitted data is confusing and self-referential. Instead, the filter should `print_where_clause` should be extracted out to a method of `clean::Generics`:

```
        let mut where_predicates = gens.where_predicates.iter().filter(|pred| {
            !matches!(pred, clean::WherePredicate::BoundPredicate { bounds, .. } if bounds.is_empty())
```

For instance it could be `fn where_predicates() -> impl Iterator<Item = WherePredicate>` on `Generics` that returns all where predicates with non-empty bounds. Then this function would have two possible rendering paths: (a) when the where clause is present (`where_predicates()` returns a nonempty iterator), and (b) when the where clause is not present.

Reply:
I agree that the code doesn't look great but i don't agree with the solution you propose. I have an idea so I'll update the PR and you can tell me if you find it better.

Reply:
I still think it's not good style. You are still checking the length of the buffer to see whether `print_where_clause` did anything: https://github.com/rust-lang/rust/pull/98256/files#diff-c4e61d56ff695cfc762517d170672a9c8650ac815abee6a9e28267a2ff27c8f3R72-R74.

That strongly suggests to me that the logic for "will print_where_clause do anything" should be extracted into its own function. I'm not particularly attached to the iterator idea I proposed above; it could also be a `has_where_clause()` function.

That will make the code that chooses whether to add a space, newline, or semicolon clearer, since it can check if there is or is not a where clause, and call `print_where_clause` only in that case; and also emit a space, newline, or semicolon only in that case.

Reply:
I don't like the idea of duplicating this logic though (duplicating in the sense of "checking it twice", not "copy/paste it in two places").